modules/auth/oauth2/oauth2.go

```diff
@@ -0,0 +1,132 @@
+package oauth2
+
+import (
+	"code.gitea.io/gitea/modules/setting"
+	"github.com/gorilla/sessions"
+	"github.com/markbates/goth"
+	"github.com/markbates/goth/gothic"
+	"github.com/markbates/goth/providers/github"
+	"gopkg.in/macaron.v1"
+	"net/http"
+	"os"
+	"github.com/go-macaron/session"
+	"github.com/satori/go.uuid"
+	"encoding/base64"
+	"strings"
+)
+
+var (
+	sessionUsersStoreKey         = "gitea-oauth-sessions"
+	oAuthUserSessionKey          = "oauth2-user"
+	oAuthStateUniqueIDSessionKey = "oauth2-state-key"
+)
+
+func init() {
+	gothic.Store = sessions.NewFilesystemStore(os.TempDir(), []byte(sessionUsersStoreKey))
+}
+
+// Auth OAuth2 auth service
+func Auth(provider, clientID, clientSecret string, ctx *macaron.Context, session session.Store) {
+	callbackURL := setting.AppURL + "user/oauth2/" + provider + "/callback"
+
+	goth.UseProviders(
+		github.New(clientID, clientSecret, callbackURL, "user:email"),
+	)
+
+	gothic.GetProviderName = func(req *http.Request) (string, error) {
+		return provider, nil
+	}
+
+	gothic.SetState = func(req *http.Request) string {
+		state, uniqueID := encodeState(req.URL.String())
+
+		session.Set(oAuthStateUniqueIDSessionKey, uniqueID)
+
+		return state
+	}
+
+	request := ctx.Req.Request
+	response := ctx.Resp
+
+	/*
```

Comment:
yes part that I still need to have a look at. Need to check if it is required to do custom storing to determine valid requests in the callback